tests/lib/schema_validation_test.py

```diff
@@ -162,3 +191,158 @@ def test_invalid_state(self, default_data):
+            1000, invalid_key='mega-boom!'
+        )
+        port_with_vlan_filter = self._get_port_vlan_filtering_empty_state(
+            'eth1'
+        )
+        self._update_port_vlan_filtering_state(
+            port_with_vlan_filter, [invalid_allow_vlan_id]
+        )
+        bridge_state = self._get_bridge_iface_state(default_data[INTERFACES])
+        bridge_state[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE] = [
+            port_with_vlan_filter
+        ]
+        with pytest.raises(js.ValidationError) as err:
+            libnmstate.validator.validate(default_data)
+        assert '(\'invalid_key\' was unexpected)' in err.value.args[0]
+
+    def test_port_vlan_filtering_invalid_range_data(self, default_data):
+        invalid_allow_vlan_range = self._get_vlan_filtering_entry_range(
+            50, 200, invalid_key='ka-boom!'
+        )
+        port_with_vlan_filter = self._get_port_vlan_filtering_empty_state(
+            'eth1'
+        )
+        self._update_port_vlan_filtering_state(
+            port_with_vlan_filter, [invalid_allow_vlan_range]
+        )
+        bridge_state = self._get_bridge_iface_state(default_data[INTERFACES])
+        bridge_state[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE] = [
+            port_with_vlan_filter
+        ]
+        with pytest.raises(js.ValidationError) as err:
+            libnmstate.validator.validate(default_data)
+        assert '(\'invalid_key\' was unexpected)' in err.value.args[0]
+
+    def test_port_vlan_filtering_valid_data(self, default_data):
+        valid_vlan_id_state = self._get_vlan_filtering_entry(1000)
+        valid_vlan_range_state = self._get_vlan_filtering_entry_range(50, 200)
+        port_with_vlan_filter = self._get_port_vlan_filtering_empty_state(
+            'eth1'
+        )
+        self._update_port_vlan_filtering_state(
+            port_with_vlan_filter,
+            [valid_vlan_id_state, valid_vlan_range_state],
+        )
+        bridge_state = self._get_bridge_iface_state(default_data[INTERFACES])
+        bridge_state[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE] = [
+            port_with_vlan_filter
+        ]
+        libnmstate.validator.validate(default_data)
+
+    def test_invalid_port_config_only_id_or_range(self, default_data):
+        invalid_port_config = {
+            'name': 'eth1',
+            'vlan-filtering': {
+                'whitelist': [{'id': 1000, 'id-range': {'min': 50, 'max': 80}}]
+            },
+        }
+        bridge_state = self._get_bridge_iface_state(default_data[INTERFACES])
+        bridge_state[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE] = [
+            invalid_port_config
+        ]
+        with pytest.raises(js.ValidationError) as err:
+            libnmstate.validator.validate(default_data)
+        assert 'is valid under each of' in err.value.args[0]
+
+    def test_invalid_linux_bridge_vlan_range(self, default_data):
+        invalid_vlan_id_type = self._get_vlan_filtering_entry('boom!')
```

Comment:
Is building these dictionaries really worth it? I find it hard to follow the code to see what is going on and to imagine what would be the result. If I saw a mere dictionary, it would be obvious right away.